File: basic_dsa_two_hundred_twenty_nine.cpp
/**
 * Basic DSA - Spiral Matrix (LeetCode 54)
 * Return elements of matrix in spiral order.
 */
#include <iostream>
#include <vector>

std::vector<int> spiralOrder(std::vector<std::vector<int>>& matrix) {
    std::vector<int> out;
    int top = 0, bot = matrix.size() - 1, left = 0, right = matrix[0].size() - 1;
    while (top <= bot && left <= right) {
        for (int j = left; j <= right; j++) out.push_back(matrix[top][j]);
        top++;
        if (top > bot) break;
        for (int i = top; i <= bot; i++) out.push_back(matrix[i][right]);
        right--;
        if (left > right) break;
        for (int j = right; j >= left; j--) out.push_back(matrix[bot][j]);
        bot--;
        if (top > bot) break;
        for (int i = bot; i >= top; i--) out.push_back(matrix[i][left]);
        left++;
    }
    return out;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Spiral Matrix\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<std::vector<int>> m1 = {{1,2,3},{4,5,6},{7,8,9}};
    auto r1 = spiralOrder(m1);
    std::cout << "[[1,2,3],[4,5,6],[7,8,9]] -> [";
    for (size_t i = 0; i < r1.size(); i++) std::cout << (i ? "," : "") << r1[i];
    std::cout << "]\n";

    std::vector<std::vector<int>> m2 = {{1,2,3,4},{5,6,7,8},{9,10,11,12}};
    auto r2 = spiralOrder(m2);
    std::cout << "[[1,2,3,4],[5,6,7,8],[9,10,11,12]] -> [";
    for (size_t i = 0; i < r2.size(); i++) std::cout << (i ? "," : "") << r2[i];
    std::cout << "]\n";
}

File: basic_dsa_two_hundred_twenty_nine.py
"""
Basic DSA - Longest Substring Without Repeating Characters
Given a string s, find the length of the longest substring without repeating characters.
"""


def length_of_longest_substring(s):
    """
    Sliding window with last seen positions.
    Time Complexity: O(n)
    Space Complexity: O(min(n, alphabet))
    """
    last = {}
    left = 0
    best = 0
    for right, ch in enumerate(s):
        if ch in last and last[ch] >= left:
            left = last[ch] + 1
        last[ch] = right
        best = max(best, right - left + 1)
    return best


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Longest Substring Without Repeating Characters")
    print("=" * 60)

    print(length_of_longest_substring("abcabcbb"))  # 3
    print(length_of_longest_substring("bbbbb"))  # 1
    print(length_of_longest_substring("pwwkew"))  # 3
